Add wrapper module for Agda pretty-printer

JonasHoefer opened the pull request:
Adds a `Pretty` instance for Agda AST nodes, as well as some (provisional) smart constructors in `FreeC.Backend.Agda.Syntax` to test the pretty printer. These constructors will probably change when work on #31 starts. Closes #29

JonasHoefer committed:
Add `module` for Agda pretty printing #29

The new module `FreeC.Backend.Agda.Pretty` contains `Pretty` instances for Agda AST nodes. The implements use Agdas internal pretty printer.

JonasHoefer committed:
Add some smart constructors for Agda AST nodes #29

This commit adds some smart constructors for Agda AST nodes to easily test the pretty printer.

JonasHoefer committed:
Add docs for (provisional) smart constructors #29

just95 commented on src/lib/FreeC/Backend/Agda/Pretty.hs:
```diff
@@ -0,0 +1,44 @@
+
+-- | This module contains 'Pretty' instances for the Agda AST.
+--
+--   The Agda has its own 'Agda.Pretty' class in 'Agda.Utils.Pretty'. Therefore
```
`Agda.Utils.Pretty` is a module. Thus, double quotes are needed in Haddock to link to it.

```suggestion
--   The Agda library has its own 'Agda.Pretty' class in "Agda.Utils.Pretty". Therefore
```

just95 commented on src/lib/FreeC/Backend/Agda/Pretty.hs:
```diff
@@ -0,0 +1,44 @@
+--   'Pretty' instance without generating overlapping instances for basic types
+--   (e.g. 'String').
+--   We introduce the wrapper datatype 'PrettyAgda' to avoid the overlapping
+--   instances and explicitly declare 'Pretty' instances for for common AST
```
```suggestion
--   instances and explicitly declare 'Pretty' instances for common AST
```

just95 commented on src/lib/FreeC/Backend/Agda/Pretty.hs:
```diff
@@ -0,0 +1,44 @@
+  )
+where
+
+-- import qualified Agda.Syntax.Concrete.Generic  as Agda
```
When commenting out code, try to use block comments to separate them from regular comments. Is there a reason why you kept this import declaration?

```suggestion
{- import qualified Agda.Syntax.Concrete.Generic  as Agda -}
```

just95 commented on src/lib/FreeC/Backend/Agda/Syntax.hs:
```diff
@@ -3,7 +3,59 @@
+  Agda.DoOpen
+  (Agda.ImportDirective Agda.NoRange Agda.UseEverything [] [] Nothing)
+
+-- | Creates a (not qualified) Agda variable name from a 'String'.
+name :: String -> Agda.Name
+name ident = Agda.Name Agda.NoRange Agda.InScope [Agda.Id ident]
```
I would move this to the other smart constructors for names and maybe add some sections like the following.

```haskell
-------------------------------------------------------------------------------
-- Names                                                                     --
-------------------------------------------------------------------------------

…

-------------------------------------------------------------------------------
-- Imports                                                                   --
-------------------------------------------------------------------------------

…

-------------------------------------------------------------------------------
-- Expressions                                                               --
-------------------------------------------------------------------------------

…
```

JonasHoefer committed:
Fix Haddock link

Co-authored-by: Justin Andresen <jan.dresen.95@gmail.com>

JonasHoefer committed:
Remove an unused import #29

JonasHoefer committed:
Add secttions to `FreeC.Backend.Agda.Syntax` #29

JonasHoefer committed:
Fix a typo in a comment #29

Co-authored-by: Justin Andresen <jan.dresen.95@gmail.com>

JonasHoefer committed:
Reword comments in `FreeC.Backend.Agda.Pretty` #29

just95 commented on src/lib/FreeC/Backend/Agda/Syntax.hs:
```diff
@@ -3,7 +3,71 @@
   ( module Agda.Syntax.Concrete
+  , module Agda.Syntax.Common
+  , module Agda.Syntax.Position
+  , name
+  , intLiteral
+  , simpleImport
+  , lambda
+  , qname
+  , qname'
```
I usually try to organize the export list in the same way I've organized the rest of the module.

```suggestion
    -- * Identifiers
  , name
  , qname
  , qname'
    -- * Imports
  , simpleImport
    -- * Expressions
  , intLiteral
  , lambda
```

just95 commented on src/lib/FreeC/Backend/Agda/Syntax.hs:
```diff
@@ -3,7 +3,71 @@
+
+-- | Create a qualified name given a local variable 'Name' and a list of module
+--   'Name's.
+qname :: Agda.Name -> [Agda.Name] -> QName
+qname n = foldr Agda.Qual (Agda.QName n)
```
Since the module name is usually written before the name, I'd swap the arguments of `qname`.

JonasHoefer committed:
Organize exports `Syntax.hs` #29

Co-authored-by: Justin Andresen <jan.dresen.95@gmail.com>

JonasHoefer committed:
Swap `qname`s arguments for readability

JonasHoefer committed:
Improve `qname`s argumet names and documentation

just95 commented on src/lib/FreeC/Backend/Agda/Syntax.hs:
```diff
@@ -3,7 +3,74 @@
+import           Agda.Syntax.Position
+import           Agda.Syntax.Common
+
+import qualified Agda.Syntax.Common            as Agda
+import qualified Agda.Syntax.Concrete          as Agda
+import qualified Agda.Syntax.Literal           as Agda
+import qualified Agda.Syntax.Position          as Agda
```
Importing the modules both qualified and unqualified has the unfortunate side effect that the `Agda` prefix can be forgotten without GHC complaining (for example, I just noticed that `QName` in `qname` is not qualified). Since it could be confusing if we mix `QName`  and `Agda.QName` but only unqualified imports can be reexported, the best option is probably to remove the qualified imports and not qualify identifiers from the Agda library in this module. I'll followup with the corresponding change to `FreeC.Backend.Coq.Syntax`.

JonasHoefer replied:
This sounds reasonable. I noticed the same thing but decided proceed in the same way as in `FreeC.Backend.Coq.Syntax`. Another advantage is that the code becomes more readable, because the smart constructors usually just provide lots of default arguments, which all had to be prefixed.

JonasHoefer committed:
Remove qualified imports from Agda's Syntax.hs #29

The module `FreeC.Backend.Agda.Syntax` imports multiple modules twice, once qualified and once unqualified. The unqualified version is needed to reexport them. This allows to use different names for the same functions (e.g. `Agda.QName` and `QName`). To avoid this we decided not to use qualified imports in this modul.
